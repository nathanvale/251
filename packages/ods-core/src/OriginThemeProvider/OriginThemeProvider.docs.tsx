import React from "react";
import { ComponentDocs } from "@origin-digital/ods-types";
import { Text } from "../Text";
import { TextLink } from "../TextLink";
import { OriginThemeProviderProps } from "./OriginThemeProvider";

export const docs: ComponentDocs<OriginThemeProviderProps> = {
  category: "Logic",
  componentName: "OriginThemeProvider",
  description:
    "OriginThemeProvider lets you apply a consistent tone to your app. It allows you to customize all design aspects of your project in order to meet the specific needs of our brands. By default the OriginThemeProvider comes with the Origin Energy retail theme.",
  propDescriptions: {
    linkComponent:
      "The custom link component you would like to use for ODS link components.",
  },
  migrationGuide: false,
  examples: {
    default: {
      codeString: `import { OriginThemeProvider } from "@origin-digital/ods-core";
export default ({ children }: { children: React.ReactChild }) => (
  <OriginThemeProvider>{children}</OriginThemeProvider>
);
      `,
    },
    additional: [
      {
        label: "Custom link implementation",
        description: (
          <Text>
            By default ODS link components such as{" "}
            <TextLink href="/components/TextLink">TextLink</TextLink>,{" "}
            <TextLink href="/components/ChevronLink">ChevronLink</TextLink> and{" "}
            <TextLink href="/components/HeadingChevronLink">
              HeadingChevronLink
            </TextLink>{" "}
            all use <code>&lt;a/&gt;</code> tags by default. If youâ€™d like to
            customise the implementation of these components (e.g. to use a
            React Router link) you can pass in your own custom link
            implementation.
          </Text>
        ),
        codeString: `
        import React from 'react';
        import { Link as ReactRouterLink } from 'react-router-dom';
        import { OriginThemeProvider } from "@origin-digital/ods-core";
        import { LinkComponentProps } from "@origin-digital/ods-types";

        // First create the custom link implementation:
        const CustomLink = React.forwardRef<any, LinkComponentProps>((props, ref) => {
          const { href, ...restProps } = props;
          return href[0] === "/" ? (
            // For tracking to be active,  
            // You must wrap with React.forardRef and pass down.
            <ReactRouterLink
            to={href}
            ref={ref}
            {...restProps}
            />
            ) : (
            <a href={href} ref={ref} {...restProps} />
          );
        });

        // Then pass it to OriginThemeProvider:
        export const App = () => (
          <OriginThemeProvider linkComponent={CustomLink}>
            ...
          </OriginThemeProvider>
        );
      `,
      },
    ],
  },
  snippets: [],
};
