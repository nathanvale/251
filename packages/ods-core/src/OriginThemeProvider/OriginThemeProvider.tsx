import React, { createContext, useContext } from "react";
import {
  ThemeProvider as MUIThemeProvider,
  Theme as MuiTheme,
} from "@material-ui/core";

import {
  LinkComponentProps,
  LinkComponentType,
} from "@origin-digital/ods-types";
import { coreMuiTheme } from "@origin-digital/ods-themes";
import { LayoutThemeProvider } from "../_private/components/LayoutThemeProvider/LayoutThemeProvider";

const DefaultLinkComponent = React.forwardRef<
  HTMLAnchorElement,
  LinkComponentProps
>((props, ref) => {
  const { children, ...restProps } = props;
  return (
    <a {...restProps} ref={ref}>
      {children}
    </a>
  );
});
const LinkComponentContext = createContext<LinkComponentType>(
  DefaultLinkComponent
);
export const useLinkComponent = () => useContext(LinkComponentContext);

export interface OriginThemeProviderProps {
  children: React.ReactNode;
  linkComponent?: LinkComponentType;
  muiTheme?: MuiTheme;
}

export const OriginThemeProvider = ({
  children,
  muiTheme = coreMuiTheme,
  linkComponent,
}: OriginThemeProviderProps) => {
  const linkComponentFromContext = useLinkComponent();
  return (
    <MUIThemeProvider theme={muiTheme}>
      <LayoutThemeProvider muiTheme={muiTheme}>
        <LinkComponentContext.Provider
          value={linkComponent || linkComponentFromContext}
        >
          {children}
        </LinkComponentContext.Provider>
      </LayoutThemeProvider>
    </MUIThemeProvider>
  );
};
