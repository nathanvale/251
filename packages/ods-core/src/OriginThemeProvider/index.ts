export { OriginThemeProvider } from "./OriginThemeProvider";
export type OriginThemeProviderProps = import("./OriginThemeProvider").OriginThemeProviderProps;
