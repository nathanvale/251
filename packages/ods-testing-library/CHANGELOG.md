# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

# [0.3.0](https://bitbucket.orgn.io/od/origin-ui/compare/@origin-digital/ods-testing-library@0.2.0...@origin-digital/ods-testing-library@0.3.0) (2020-02-12)


### Features

* **tests, columns:** alignY default changed to top ([c23c964](https://bitbucket.orgn.io/od/origin-ui/commits/c23c964))





# 0.2.0 (2020-01-31)


### Features

* set up jest for snapshots ([6e6fbfa](https://bitbucket.orgn.io/od/origin-ui/commits/6e6fbfa))
