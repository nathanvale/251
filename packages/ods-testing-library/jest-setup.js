import { configure } from "@testing-library/react";

require("jest-styled-components");

configure({ testIdAttribute: "data-id" });
