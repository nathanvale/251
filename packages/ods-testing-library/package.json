{
  "name": "@origin-digital/ods-testing-library",
  "description": "ODS testing tools that enhance react-testing-library",
  "license": "UNLICENSED",
  "author": "Nathan Vale <nathan.vale@origin.com.au>",
  "version": "0.5.0",
  "main": "dist/cjs/index.js",
  "module": "dist/es/index.js",
  "files": [
    "dist",
    "jest-setup.js"
  ],
  "scripts": {
    "build": "yarn build:cjs && yarn build:es",
    "build:cjs": "BUILD_FORMAT=cjs origin-scripts build --monorepo --delete-dir-on-start --out-dir dist/cjs",
    "build:es": "BUILD_WEBPACK=true origin-scripts build --monorepo --out-dir dist/es"
  },
  "sideEffects": false,
  "typings": "dist/typings/index.d.ts",
  "dependencies": {
    "@babel/runtime-corejs3": "^7.8.3",
    "@testing-library/jest-dom": "^5.1.1"
  },
  "peerDependencies": {
    "@origin-digital/ods-themes": "*",
    "@testing-library/react": "9.x",
    "jest-styled-components": "6.x",
    "react": "^16.9.0"
  },
  "devDependencies": {
    "@testing-library/react": "9.x"
  },
  "publishConfig": {
    "registry": "https://artifactory.origindigital-dac.com.au/artifactory/api/npm/npm-internal/"
  }
}
