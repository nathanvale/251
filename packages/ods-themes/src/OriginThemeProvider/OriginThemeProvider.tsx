import React from "react";
import { ThemeProvider, injectGlobal } from "styled-components";
import MUIThemeProvider from "@material-ui/styles/ThemeProvider";

import createMuiTheme from "@material-ui/core/styles/createMuiTheme";

import { Theme } from "@origin-digital/ods-types";
import { originRetailTheme } from "../themes";
import { originRetailMUITheme } from "../_private/mui-themes/originRetail";

export interface OriginThemeProviderProps {
  theme?: Theme;
  children: React.ReactNode;
}

// eslint-disable-next-line babel/no-unused-expressions
injectGlobal`
      html {
        background-color: #f1f1f1;
        height:100%;
      }

      body {
        margin: 0;
        font-family: gordita, sans-serif;
        height: 100%;
      }

      body > div {
        height: 100%;
      }

      @font-face {
        font-family: 'gordita';
        src: url(https://origin-fonts.s3-ap-southeast-2.amazonaws.com/Origin_Gordita-Light.woff2)
          format('woff');
        font-weight: 200;
        font-style: normal;
      }

      @font-face {
        font-family: 'gordita';
        src: url(https://origin-fonts.s3-ap-southeast-2.amazonaws.com/Origin_Gordita-Regular.woff2)
          format('woff');
        font-weight: 500;
        font-style: normal;
      }

      @font-face {
        font-family: 'gordita';
        src: url(https://origin-fonts.s3-ap-southeast-2.amazonaws.com/Origin_Gordita-Medium.woff2)
          format('woff');
        font-weight: 600;
        font-style: normal;
      }

      @font-face {
        font-family: 'gordita';
        src: url(https://origin-fonts.s3-ap-southeast-2.amazonaws.com/Origin_Gordita-Bold.woff2)
          format('woff');
        font-weight: 900;
        font-style: normal;
      }
`;

export const OriginThemeProvider = ({
  children,
  theme = originRetailTheme,
}: OriginThemeProviderProps) => {
  //TODO: create a mapper function that maps originRetailTheme to a originRetailMUITheme
  // https://origindd.atlassian.net/browse/TED-484
  const muiTheme = createMuiTheme(originRetailMUITheme as any);
  return (
    <ThemeProvider theme={theme}>
      <MUIThemeProvider theme={muiTheme}>{children}</MUIThemeProvider>
    </ThemeProvider>
  );
};
