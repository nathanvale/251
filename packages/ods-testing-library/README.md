# @origin-digital/ods-testing-library

<!-- START doctoc generated TOC please keep comment here to allow auto update -->
<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->

- [Installation](#installation)
- [Why do I need this package?](#why-do-i-need-this-package)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

This package enhances the render function of `react-testing-library` so that the `<OriginThemeProvider/>` is wrapped around your rendered component.

All other API features of `react-testing-library` are also included.

## Installation

Install the package in your project directory with:

```sh
yarn add @origin-digital/ods-testing-library
```

This package also has a peer dependency on the following:

```sh
// with yarn
yarn add @origin-digital/ods-core
yarn add -D @testing-library/react
yarn add -D jest-styled-components
```

## Why do I need this package?

**The problem**

When using react-testing-library and ODS with the jest testing framework it is necessary to wrap **all** your components in our `<OriginThemeProvider/>` - otherwise the react component cannot be rendered and tested.

**The solution**

`@origin-digital/ods-testing-library` creates a wrapper around the `react-testing-library` render function. It is here that we wrap the `<OriginThemeProvider/>`. All other API features of `react-testing-library` are also included.

This practice is recommended by the `react-testing-library` documentation itself...

https://testing-library.com/docs/native-testing-library/setup#custom-render

The origin-ui dogfoods this package itself but it can also be used in your own app to test by adding to your jest config:

```
{
...,
setupFilesAfterEnv: [@origin-digital/ods-testing-library/jest-setup],
...
}
```

and usage in your tests:

```
import React from 'react';
import {render} from '@origin-digital/ods-testing-library';
import {Box} from '../Box/Box';
import {Stack} from './Stack';

test('It can stack', () => {
  const {container} = render(
    <Stack>
      <Box backgroundColor="red" padding="medium" />
      <Box backgroundColor="orange" padding="medium" />
      <Box backgroundColor="blue" padding="medium" />
    </Stack>,
  );
  expect(container.firstChild).toMatchSnapshot();
});
```
