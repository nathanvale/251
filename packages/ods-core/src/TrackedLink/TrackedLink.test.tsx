import React from "react";
import { render, fireEvent } from "@origin-digital/ods-testing-library";
import { LinkComponentType } from "@origin-digital/ods-types";
import { OriginThemeProvider } from "../OriginThemeProvider/OriginThemeProvider";
import { TrackedLink, TrackingProvider, Box } from "..";

test("It can render", () => {
  const { container } = render(
    <TrackedLink trackingType="CustomLink" trackingLabel="Tracked Link" href="">
      Tracked Link
    </TrackedLink>
  );
  expect(container.firstChild).toMatchSnapshot();
});

test("it handles a custom link implementation", () => {
  const onTrackingCapture = jest.fn();
  const onClick = jest.fn();

  const CustomLink: LinkComponentType = ({ href, ...restProps }) => (
    // eslint-disable-next-line jsx-a11y/anchor-has-content
    <a href={href} onClick={onClick} {...restProps} />
  );

  const { getByTestId } = render(
    <OriginThemeProvider linkComponent={CustomLink}>
      <TrackingProvider onTrackingCapture={onTrackingCapture}>
        <TrackedLink
          href=""
          trackingType="CustomLink"
          trackingLabel="The big red box is being tracked!"
          data-id="custom-link"
        >
          <Box backgroundColor="critical" padding="large" />
        </TrackedLink>
      </TrackingProvider>
    </OriginThemeProvider>
  );

  const link = getByTestId("custom-link");
  fireEvent.click(link);

  expect(onClick).toBeCalledTimes(1);
  expect(onTrackingCapture).toBeCalledWith({
    "data-id": "custom-link",
    label: "The big red box is being tracked!",
    postClickState: undefined,
    type: "CustomLink",
  });
});
