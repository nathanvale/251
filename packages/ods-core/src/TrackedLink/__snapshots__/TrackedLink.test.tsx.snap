// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`It can render 1`] = `
<a
  data-id="tracked-link"
  href=""
>
  Tracked Link
</a>
`;
